trading/trading.go

```diff
@@ -42,27 +47,71 @@ func (p Portfolio) Value(prices Prices) float64 {
 	return value
 }
 
-// Abstract portfolio. Can hold fractional and negative shares.
+// Abstract portfolio. A probability distribution over potential assets.
 type AbstractPortfolio struct {
-	Stocks  map[string]float64
-	CashUSD float64
+	stocks map[string]float64
+	total  float64
 }
 
-func NewAbstractPortfolio(seed float64) AbstractPortfolio {
-	return AbstractPortfolio{
-		Stocks:  map[string]float64{},
-		CashUSD: seed,
+func NewAbstractPortfolio() AbstractPortfolio {
+	return AbstractPortfolio{stocks: map[string]float64{}, total: 0.0}
+}
+
+func NewBalancedAbstractPortfolio(stocksPrices Prices) {
+	portfolio := NewAbstractPortfolio()
+	for symbol, _ := range prices {
+		portfolio.SetStock(symbol, 1.0)
 	}
+	return portfolio
+}
 
+// Ensures that the abstract portfolio is normalized to be a distribution.
+func (a *AbstractPortfolio) ensureDistribution() {
+	if a.total == 0.0 {
+		log.Fatalf("Tried to create distribution in AbstractPortfolio before SetStock.")
+	} else if a.total == 1.0 {
+		return
+	}
+	for symbol, value := range a.stocks {
+		a.stocks[symbol] = value / a.total
+	}
+	a.total = 1.0
 }
 
-// Returns mature value of a position given a Tick.
-func (p AbstractPortfolio) Value(prices Prices) float64 {
-	value := p.CashUSD
-	for symbol, quantity := range p.Stocks {
-		value += quantity * prices[symbol]
+// Sets a value for stock and updates the total.
+func (a *AbstractPortfolio) SetStock(symbol string, value float64) {
+	a.total -= a.stocks[symbol]
+	a.stocks[symbol] = value
+	a.total += value
+}
+
+// Gets stock (after ensuring the portfolio is a distribution).
+func (a *AbstractPortfolio) GetStock(symbol string) float64 {
+	a.ensureDistribution()
```

Comment:
I guess since I have the early return in ensureDistribution it won't ultimately matter much.

The thinking was: add a bunch of stocks but only ensureDistribution once. But I realize that's not well founded.. you could just as easily have a case where you add a few stocks and read many more times than that.